### ISSUE TITLE
request_strategy parameter for APIClient constructor

### ISSUE THREAD

**Issue by peterschutt** [MikeWooster/api-client#58]
Hi Mike, 

I'm enjoying using this library, so thanks for your time and efforts putting it together!

**Is your feature request related to a problem? Please describe.**
I'd like to pass a `RequestStrategy` to the client constructor.

**Describe the solution you'd like**
Similar to how the client accepts `authentication_method` et al. as constructor params, I'd like to be able to pass a `request_strategy` instance one way or another.

**Describe alternatives you've considered**
I construct the client and then call `set_request_strategy()` on the client instance.

**Additional context**
Is it that the request strategy needs to be a distinct instance per client given the client is persisted on the strategy? Would you consider a parameter that accepts a callable which returns a strategy instance? It might be `RequestStrategy` type or something more complicated. Or perhaps a `None` default value for parameter that accepts an instance so that if nothing is passed in a fresh `RequestStragegy` can be constructed for the client?

As mentioned above, in terms of LOC this is only saving 1 call to `set_request_strategy()` so I'd entirely understand if you decide its not worth the effort!

**Comment by mom1** [comment]
I would like to suggest setting this parameter in the attribute of the child class. so as not to override the __init__ method.

**Comment by peterschutt** [comment]
@mom1 Would you mind demonstrating your suggestion for me? I'm not sure that I understand.

However, my goal is to raise a different exception from the request other than the `APIRequestException` type (specifically a `fastapi.HTTPException`), and I've just come across the PR #56 that you are working on which seems it might give me a nice clean interface to manage that anyway, rather than overriding `_make_request()` on the strategy.

**Comment by peterschutt** [comment]
@mom1 Thanks for updating with the example. How should the client instance get passed to the strategy in such a scenario?

**Comment by mom1** [comment]
In my case, ApiClient uses MyRequestStrategy more often. 
I am now forced to override `__init__` to set the default for request_strategy, or call set_request_strategy () all the time.
And for change I would like to use MyClient (request_strategy = OtherStrategy ())

**Comment by peterschutt** [comment]
Another option could be a `get_request_strategy()` method on the `APIClient` so that it could be easily overridden:

```python
class APIClient:

    def __init__(self, ...) -> None:
        ...
        self.set_request_strategy(self.get_request_strategy())
        ...

    def get_request_strategy(self) -> RequestStrategy:
        return RequestStrategy()

```

**Comment by mom1** [comment]
@peterschutt Cool. Maybe even so.
```python
class APIClient:
    _request_strategy = RequestStrategy()
    
    def __init__(self, ...) -> None:
        ...
        self.set_request_strategy(self.get_request_strategy())
        ...

    def get_request_strategy(self) -> BaseRequestStrategy:
        return self._request_strategy
  ```

**Comment by MikeWooster** [comment]
Many thanks for your message.  I'm glad that people are using this and are actually finding it useful.  As you can probably appreciate, I wrote this to satisfy my own personal requirements. And it's good to see how others are adapting it/using it to meet their own.

I'd rather not use the class variable approach, as it is not one that appears in the rest of the client, so would be inconsistent with the way the rest works.

I would however be happy with adding a new parameter to the `__init__` method of the client so that you can pass in your custom strategy when you want. It can default to the current one so peoples current implementations continue to function.

**Comment by mom1** [comment]
> I'd rather not use the class variable approach, as it is not one that appears in the rest of the client, so would be inconsistent with the way the rest works.

I think rest of the client use `self.get_request_strategy()` this means it will work.

And this will allow instead of such classes:
```python
class MyAPIClient(APIClient):
    def __init__(
        self,
        authentication_method: BaseAuthenticationMethod = NoAuthentication(),
        response_handler: Type[BaseResponseHandler] = JsonResponseHandler,
        request_formatter: Type[BaseRequestFormatter] = MyJsonRequestFormatter,
        request_strategy: BaseRequestStrategy = MyRequestStrategy()
    ):
        super().__init__(
            authentication_method=authentication_method,
            response_handler=response_handler,
            request_formatter=request_formatter,
            request_strategy=request_strategy,
        )
    ...
```

Will allow create classes simpler and more understandable. If we use this approach for all parameters.

```python
class MyAPIClient(APIClient):
    authentication_method: BaseAuthenticationMethod = NoAuthentication(),
    response_handler: Type[BaseResponseHandler] = JsonResponseHandler,
    request_formatter: Type[BaseRequestFormatter] = MyJsonRequestFormatter,
    request_strategy: BaseRequestStrategy = MyRequestStrategy()
```

**Comment by MikeWooster** [comment]
> > I'd rather not use the class variable approach, as it is not one that appears in the rest of the client, so would be inconsistent with the way the rest works.
> 
> I think rest of the client use `self.get_request_strategy()` this means it will work.
> 
> And this will allow instead of such classes:
> 
> ```python
> class MyAPIClient(APIClient):
>     def __init__(
>         self,
>         authentication_method: BaseAuthenticationMethod = NoAuthentication(),
>         response_handler: Type[BaseResponseHandler] = JsonResponseHandler,
>         request_formatter: Type[BaseRequestFormatter] = MyJsonRequestFormatter,
>         request_strategy: BaseRequestStrategy = MyRequestStrategy()
>     ):
>         super().__init__(
>             authentication_method=authentication_method,
>             response_handler=response_handler,
>             request_formatter=request_formatter,
>             request_strategy=request_strategy,
>         )
>     ...
> ```

@mom1 could you explain your reason for wanting to do this, other than wanting to provide a different `request_strategy`?  If you have further init to perform, perhaps we could provide a `pre_init` and `post_init` methods that you could override.  that would make it slightly cleaner.

> Will allow create classes simpler and more understandable. If we use this approach for all parameters.
> 
> ```python
> class MyAPIClient(APIClient):
>     authentication_method: BaseAuthenticationMethod = NoAuthentication(),
>     response_handler: Type[BaseResponseHandler] = JsonResponseHandler,
>     request_formatter: Type[BaseRequestFormatter] = MyJsonRequestFormatter,
>     request_strategy: BaseRequestStrategy = MyRequestStrategy()
> ```

From my experience, doing this means you are making testing more difficult for yourself. I think being able to pass in the strategy makes sense; you could then pass in a test strategy instead to allow you to test your client without making any calls.  When they are class variables, they get loaded at runtime, which makes them more difficult to patch.

**Comment by peterschutt** [comment]
I've submitted a PR with the simple change of allowing a `request_strategy` parameter to the `APIClient` constructor that preserves the current behavior, however, happy to make any changes based on discussion here.

**Comment by mom1** [comment]
> @mom1 could you explain your reason for wanting to do this, other than wanting to provide a different `request_strategy`? 

1. I have my custom strategies and I do not want all in my code write `MyClient(request_strategy= MyRequestStrategy())` I want write `MyClient()`
2. `super().__init__(...)` This code needs testing. the class variable no need testing.
3. My client became clear and short.
4. I do not suggest removing the possibility of overriding through `__init__`.
I suggest something like this
```python
class APIClient:
    request_strategy_cls = RequestStrategy
    def __init__(
        self,
        ...,
        request_strategy: Optional[BaseRequestStrategy] = None,
    ):
        ...
        self.set_request_strategy(request_strategy)

        def set_request_strategy(self, request_strategy: Optional[BaseRequestStrategy] = None):
            
            request_strategy = request_strategy or self.request_strategy_cls()
            
            if not isinstance(request_strategy, BaseRequestStrategy):
                raise RuntimeError("provided request_strategy must be an instance of BaseRequestStrategy.")
            self._request_strategy = request_strategy
            self._request_strategy.set_client(self)
```

Additionally, this gives me opportunity set my strategy:
  - `MyClient(request_strategy= MyRequestStrategy())` 
  - `MyClient().set_request_strategy(MyRequestStrategy())`

And reset to default strategy

```python
client = MyClient(request_strategy= MyRequestStrategy())
...
client.set_request_strategy()
```

**Comment by MikeWooster** [comment]
resolved by #58
